catalyst-gateway-crates/c509-certificate/src/c509_general_names/general_name.rs

```diff
@@ -0,0 +1,346 @@
+use crate::c509_oid::C509oid;
+
+/// A struct represents a `GeneralName`.
+/// ```cddl
+/// GeneralName = ( GeneralNameType : int, GeneralNameValue : any )
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct GeneralName {
+    /// An enum of registered `GeneralName`.
+    gn: GeneralNameRegistry,
+    /// The value of the `GeneralName` in `GeneralNameValue`.
+    value: GeneralNameValue,
+}
+
+#[allow(dead_code)]
+impl GeneralName {
+    /// Create a new instance of `GeneralName`.
+    #[must_use]
+    pub fn new(gn: GeneralNameRegistry, value: GeneralNameValue) -> Self {
+        Self { gn, value }
+    }
+
+    /// Get the `GeneralName` type.
+    #[must_use]
+    pub fn get_gn(&self) -> &GeneralNameRegistry {
+        &self.gn
+    }
+
+    /// Get the value of the `GeneralName` in `GeneralNameValue`.
+    #[must_use]
+    pub fn get_gn_value(&self) -> &GeneralNameValue {
+        &self.value
+    }
+}
+
+impl Encode<()> for GeneralName {
+    fn encode<W: Write>(
+        &self, e: &mut Encoder<W>, ctx: &mut (),
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        // Encode GeneralNameType as int
+        let i = self
+            .gn
+            .get_int()
+            .map_err(minicbor::encode::Error::message)?;
+        e.i16(i)?;
+        // Encode GeneralNameValue as its type
+        self.value.encode(e, ctx)?;
+        Ok(())
+    }
+}
+
+impl Decode<'_, ()> for GeneralName {
+    fn decode(d: &mut Decoder<'_>, _ctx: &mut ()) -> Result<Self, minicbor::decode::Error> {
+        if minicbor::data::Type::U8 == d.datatype()? || minicbor::data::Type::I8 == d.datatype()? {
+            let i = d.i16()?;
+            // Get the name from the int value
+            let gn = *GENERAL_NAME_TABLES
+                .get_int_to_name_table()
+                .get_map()
+                .get_by_left(&i)
+                .ok_or(minicbor::decode::Error::message(
+                    "GeneralName int value not found in the table",
+                ))?;
+            // Get the value type from the int value
+            let value_type = *GENERAL_NAME_TABLES
+                .get_int_to_type_table()
+                .get(&i)
+                .ok_or_else(|| {
+                    minicbor::decode::Error::message("Extension value type not found")
+                })?;
+            Ok(GeneralName::new(
+                gn,
+                GeneralNameValue::decode(d, &mut value_type.get_type())?,
+            ))
+        } else {
+            // GeneralName is not type int
+            Err(minicbor::decode::Error::message(
+                "GeneralName id type invalid, expected int",
+            ))
+        }
+    }
+}
+
+// -----------------GeneralNameRegistry------------------------
+
+/// Enum of `GeneralName` registered in table Section 9.9 C509.
+#[allow(clippy::module_name_repetitions)]
+#[derive(Debug, Copy, PartialEq, Clone, Eq, Hash, EnumIs)]
+pub enum GeneralNameRegistry {
```

Comment:
```suggestion
pub enum GeneralNameTypes {
```

Reply:
Will change to `GeneralNameTypeRegistry`